gdal/frmts/mrf/mrf_band.cpp

```diff
@@ -752,7 +938,50 @@ CPLErr MRFRasterBand::IReadBlock(int xblk, int yblk, void *buffer) {
         }
     }
 
-    src.buffer = (char *)data;
+#if defined(ZSTD_SUPPORT)
+    // undo ZSTD
+    if (dozstd) {
+        if (img.pageSizeBytes > INT_MAX - 1440) {
+            CPLFree(data);
+            CPLError(CE_Failure, CPLE_AppDefined, "Page is too large at %d", img.pageSizeBytes);
+            return CE_Failure;
+        }
+        dst.size = img.pageSizeBytes + 1440; // Allow for a slight increase from previous compressions
+        dst.buffer = static_cast<char*>(VSIMalloc(dst.size));
+        if (nullptr == dst.buffer) {
+            CPLFree(data);
+            CPLError(CE_Failure, CPLE_OutOfMemory, "Cannot allocate %d bytes", static_cast<int>(dst.size));
+            return CE_Failure;
+        }
+
+        if (!poDS->pzsdctx)
+            poDS->pzsdctx = ZSTD_createDCtx();
+        auto ctx = static_cast<ZSTD_DCtx*>(poDS->pzsdctx);
+        auto raw_size = ctx ? ZSTD_decompressDCtx(ctx, dst.buffer, dst.size, src.buffer, src.size)
```

Comment:
I guess ctx == NULL should be caught as an error instead of trying ZSTD_decompress()